photon-lib/src/main/java/org/photonvision/simulation/VisionSystemSim.java

```diff
@@ -0,0 +1,379 @@
+    public void resetRobotPose(Pose2d robotPose) {
+        resetRobotPose(new Pose3d(robotPose));
+    }
+
+    /** Clears all previous robot poses and sets robotPose at current time. */
+    public void resetRobotPose(Pose3d robotPose) {
+        robotPoseBuffer.clear();
+        robotPoseBuffer.addSample(Timer.getFPGATimestamp(), robotPose);
+    }
+
+    public Field2d getDebugField() {
+        return dbgField;
+    }
+
+    /**
+     * Periodic update. Ensure this is called repeatedly-- camera performance is used to automatically
+     * determine if a new frame should be submitted.
+     *
+     * @param robotPoseMeters The current robot pose in meters
+     */
+    public void update(Pose2d robotPoseMeters) {
+        update(new Pose3d(robotPoseMeters));
+    }
+
+    /**
+     * Periodic update. Ensure this is called repeatedly-- camera performance is used to automatically
+     * determine if a new frame should be submitted.
+     *
+     * @param robotPoseMeters The current robot pose in meters
+     */
+    public void update(Pose3d robotPoseMeters) {
+        var targetTypes = targetSets.entrySet();
+        // update vision targets on field
+        targetTypes.forEach(
+                entry ->
+                        dbgField
+                                .getObject(entry.getKey())
+                                .setPoses(
+                                        entry.getValue().stream()
+                                                .map(t -> t.getPose().toPose2d())
+                                                .collect(Collectors.toList())));
+
+        if (robotPoseMeters == null) return;
+
+        // save "real" robot poses over time
+        double now = Timer.getFPGATimestamp();
+        robotPoseBuffer.addSample(now, robotPoseMeters);
+        dbgField.setRobotPose(robotPoseMeters.toPose2d());
+
+        var allTargets = new ArrayList<VisionTargetSim>();
+        targetTypes.forEach((entry) -> allTargets.addAll(entry.getValue()));
+        var visibleTargets = new ArrayList<Pose3d>();
+        var cameraPose2ds = new ArrayList<Pose2d>();
+        // process each camera
+        for (var camSim : camSimMap.values()) {
+            // check if this camera is ready to process and get latency
+            var optTimestamp = camSim.consumeNextEntryTime();
+            if (optTimestamp.isEmpty()) continue;
+            // when this result "was" read by NT
+            long timestampNT = optTimestamp.get();
+            // this result's processing latency in milliseconds
+            double latencyMillis = camSim.prop.estLatencyMs();
+            // the image capture timestamp in seconds of this result
+            double timestampCapture = timestampNT / 1e6 - latencyMillis / 1e3;
+
+            // use camera pose from the image capture timestamp
+            Pose3d lateRobotPose = getRobotPose(timestampCapture);
+            Pose3d lateCameraPose = lateRobotPose.plus(getRobotToCamera(camSim).get());
```

Comment:
getRobotToCamera returns optional, camSim confirmed to be non-empty?

Reply:
When adding a camera sim to the vision system sim, it is ensured that there is a transform. The only way for the `getRobotToCamera` method to return empty should be if the given camera sim does not exist in the vision system sim. In this case, we are iterating through the camera sims in this vision system sim.